src/token_adapter/jwt.ts

```
@@ -1,57 +1,125 @@
 import {ITokenData, TokenType} from '@essential-projects/core_contracts';
 import {ITokenAdapter} from '@essential-projects/iam_contracts';
-import {sign, SignOptions, verify} from 'jsonwebtoken';
+import {IPkiService} from '@essential-projects/pki_service_contracts';
 
-import * as BluebirdPromise from 'bluebird';
-
-const secret = 'secret';
+import {sign, SignOptions, verify, VerifyOptions} from 'jsonwebtoken';
 
 export class JwtTokenAdapter implements ITokenAdapter {
 
+  private _pkiService: IPkiService = undefined;
+  public config: any = undefined;
+
+  constructor(pkiService: IPkiService) {
+    this._pkiService = pkiService;
+  }
+
   public get tokenType(): TokenType {
     return TokenType.jwt;
   }
 
+  private get pkiService(): IPkiService {
+    return this._pkiService;
+  }
+
   public async encode(payload: ITokenData): Promise<string> {
 
-    const options: SignOptions = {};
-    const jwt = await this._sign(payload, secret, options);
-    return <string> jwt;
+    const algorithm: string = this.config && this.config.algorithm ? this.config.algorithm : 'HS256';
+    const expiresIn: string = this.config && this.config.expiresIn ? this.config.expiresIn : undefined;
+
+    const secretOrPrivateKey: string = await this._getSecretOrPrivateKey(algorithm);
+    const options: SignOptions = {
+      algorithm: algorithm,
+    };
+
+    if (expiresIn !== undefined) {
+      options.expiresIn = expiresIn;
+    }
+
+    const jwt: string = await this._sign(payload, secretOrPrivateKey, options);
+
+    return jwt;
   }
 
   public async decode(token: string): Promise<ITokenData> {
-    const data = await this._decode(token);
-    return <ITokenData> data;
+    if (token === undefined) {
+      return undefined;
```

Why fail silently here, when you could scream at the user what went wrong:

```TypeScript
if (token === undefined) {
  throw new Error(`cant decode 'undefined', because it is not a valid token!`)
}
```

---

Das hätte ebenfalls Konsequenzen bzgl. Abwärtskompatibilität. Der Fehler müsste dann woanders abgefangen werden, damit guest User möglich sein (Token wird nicht übergeben).

---

@ppoetschke breaking changes bringen nunmal konsequenzen bezüglich der Abwärtskompatibilität mit sich. Das Paket bekommt dann einfach ne entsprechende Version, und gut ist. Wer das feature mit der config nutzen will, muss doch eh seine installation anpassen

---

@heikomat Genau das stand eben nicht im Fokus. Der Aufwand erhöht sich deutlich, weitere Module sind betroffen, die PRs durchlaufen müssen. Die Tests müssen erweitert werden. Der Aufwand verdoppelt sich, was mit den Stake Holdern (HSM, Vorstart) erst besprochen werden müsste. Andererseits sollte das Problem zügig gefixt werden.
Wenn sich der Aufwand erhöht bzw. sich das Thema verzögert, müsste ich jetzt auch wieder ein Zwischenrelease auf einem FB einschieben, ergo noch mehr Aufwand.

---

@ppoetschke alternativ haben wir in Zukunft den Aufwand, weil das eh gemacht werden muss, nur mit dem unterschied, dass wir bis dahin Gefahr laufen Fehler nicht zu erkennen, weil sie einfach verschluckt werden, was wiederum zu erheblichem Debugging-Aufwand führen kann.

Heute ist es vlt schneller, den Fehler zu verschlucken, aber insgesamt ist das teurer. Das denke ich allerdings keine Diskussion, die in diesem PR geführt werden sollte.

---

@heikomat Ich denke ich habe einen anderen Weg gefunden, das im iam-Modul zu regeln, ohne einen breaking change zu produzieren.

---

@ppoetschke Warum nicht - wie wir mMn besprochen hatten - eine eigene Implementierung für Vorstart/HSM zu machen? Das ist echt nicht viel code.

Ich werde das so nur ungern in die essential-projects übernehmen, da es sich hier nur um eine Beispielimplementierung des Interface handelt.

---

@cmg-dev Ist doch schon entschieden. Siehe letzten Commit. Auch die Beispielimplementierung sollte "sauber" sein und reicht z.B. für Vorstart vollkommen aus.
Eine eigene Implementierung würde ich erst wie besprochen mit Anbindung eines 3rd Party Identity-Systems machen.